text/0192-remove-constructs-compat.md

```diff
@@ -0,0 +1,667 @@
+- [x] Branch to 4.x and release as @next
+- [x] Reintroduce `c.node` instead of `Node.of(c)`
+- [x] Removal of `onPrepare` and `onSynthesize` and all synthesis-related code.
+- [x] Reintroduce dependencies
+- [x] Change `node.dependencies` to return the list of node dependency (non recursive) and add `node.depgraph` which returns a `Graph` object from cdk8s.
+- [x] Stack trace control
+
+CDK changes:
+
+- [x] cfn2ts
+- [ ] Consider aliasing in TypeScript to reduce potential merge conflicts.
+
+- [ ] Require `app` when defining a `Stack`.
+- [ ] assets/compat.ts
+
+
+
+
+
+## Removal of "prepare"
+
+The "prepare" hook was removed from constructs since it is a very fragile API.
+Since the tree can be mutated during prepare, the order of `prepare` invocations
+becomes critical, and almost impossible to get right without a rich model of
+relationships between these "prepare" calls.
+
+The prepare hook was used in the CDK in a few cases:
+1. Resolution of references across stacks and nested stacks
+2. Resolution of dependencies between constructs across stacks
+3. Calculation of logical IDs based on hashes of other resources (API GW
+   Deployment, Lambda Version).
+
+The first two use cases have already been addressed by centralizing the "prepare" logic at the stage level (into [prepare-app.ts](TODO)).
+
+The 3rd item can be addressed using `Lazy` tokens (see
+[example](https://github.com/aws/aws-cdk/pull/8962/files#diff-51d435d71a31c2607f923fc4d96cac56R140)),
+and will be addressed on 1.x prior to the 2.x fork.
+
+## Validation changes
+
+Since construct validation is quite rare, in constructs 4.x, the `validate()` protected method was removed to clean up the namespace. The alternative is to implement an interface `IValidation` with a similar method.
+
+The breaking change is that users who implemented `validate()` would need to
+implement `IValidation` and change their `validate()` method from `protected` to
+`public`.
+
+## Stack trace settings
+
+Since stack traces are not attached to metadata entries by default in constructs
+4.x, we will need to pass `stackTrace: true` for `CfnResource`s. This will
+preserve the deploy-time stack traces which are very important for users.
+
+Other metadata entries will not get stack traces by default, and that's a
+reasonable change.
+
+When stack traces are disabled (either through the CDK context or through
+`CDK_STACK_TRACE_DISABLE`), we will need to set the appropriate context key in
+`App` so that this will propagate to "constructs".
+
+## Info/warning/error message metadata key changes
+
+The construct metadata keys for `addInfo()`, `addWarning()` and `addError()` are
+`aws:cdk:info`, etc. These are not the keys used by default in "constructs"
+since the library is not part of the AWS CDK.
+
+To address this, the "constructs" library allows customizing these keys through a settings module. We will need to add that configuration to the `App` level so this behavior will be preserved.
```

Comment:
Makes sense. And addresses https://github.com/aws/aws-cdk-rfcs/pull/195#issuecomment-656611597

Reply:
I actually decided that I agree with you that `addWarning/error/...` should not be part of `constructs` and moved them to `Logging.of(x).addWarning(msg)`.